src/Microsoft.Tye.Hosting/WatchBuilderWorker.cs

```diff
@@ -0,0 +1,180 @@
+
+        public WatchBuilderWorker(ILogger logger) 
+        {
+            _logger = logger; 
+            _queue = Channel.CreateUnbounded<BuildRequest>(); 
+            _processor = Task.Run(ProcessTaskQueueAsync);
+        }
+
+        class BuildRequest
+        {
+            public BuildRequest(string projectFilePath, string workingDirectory)
+            {
+                this.projectFilePath = projectFilePath;
+                this.workingDirectory = workingDirectory;
+            }
+
+            public string projectFilePath { get; set; }
+
+            public string workingDirectory { get; set; }
+
+            private TaskCompletionSource<int> _result = new TaskCompletionSource<int>();
+
+            public Task<int> task()
+            {
+                return _result.Task;
+            }
+
+            public void complete(int exitCode)
+            {
+                if(!_result.TrySetResult(exitCode))
+                    throw new Exception("failed to set result");
+            }
+        }
+
+        public Task<int> buildProjectFileAsync(string projectFilePath, string workingDirectory) {
+            var buildRequest = new BuildRequest(projectFilePath, workingDirectory);
+            _queue.Writer.WriteAsync(buildRequest);
+            return buildRequest.task();
+        }
+
+        public Task<int> buildProjectFileAsyncImpl(string projectFilePath, string workingDirectory) {
+            _logger.LogInformation($"Building project ${projectFilePath}...");
+            return ProcessUtil.RunAsync("dotnet", $"build \"{projectFilePath}\" /nologo", throwOnError: false, workingDirectory: workingDirectory)
+                .ContinueWith((processTask) => {
+                    var buildResult = processTask.Result;
+                    if (buildResult.ExitCode != 0)
+                    {
+                        _logger.LogInformation("Building projects failed with exit code {ExitCode}: \r\n" + buildResult.StandardOutput, buildResult.ExitCode);
+                    }
+                    return buildResult.ExitCode;
+                });
+        }
+
+        private string GetProjectName(SolutionFile solution, string projectFile)
+        {
+            foreach(var project in solution.ProjectsInOrder) {
+                if(project.AbsolutePath == projectFile)
+                {
+                    return project.ProjectName;
+                }
+            }
+            // TODO: error
+            return "";
+        }
+
+        private async Task ProcessTaskQueueAsync()
+        {
+            try
+            {
+                while (await _queue.Reader.WaitToReadAsync())
+                {
+                    var solutionBatch = new Dictionary<string, List<BuildRequest>>(); //  store the list of promises
+                    var projectBatch = new Dictionary<string, List<BuildRequest>>();
+                    // TODO: quiet time... maybe wait both...?
+                    await Task.Delay(100);
```

Comment:
Why is there a delay here?

Reply:
I had the idea of having a quiet time where we wait to see if more files will be saved, then schedule the builds at once... it would be better to have it waiting 100ms (or a more sensible duration) after the last WaitToReadAsync(). Unsure what would happen if it was removed; might still need to yield to let duplicate project build requests enter the queue?